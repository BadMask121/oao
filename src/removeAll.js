// @flow

import { set as timmSet, omit } from 'timm';
import { readAllSpecs } from './utils/readSpecs';
import { DEP_TYPES } from './utils/constants';
import { parseDep } from './utils/helpers';
import writeSpecs from './utils/writeSpecs';

type Options = {
  src: string,
  ignoreSrc?: string,
  link: ?string,
};

const run = async (deps: Array<string>, opts: Options) => {
  const { src, ignoreSrc } = opts;
  const allSpecs = await readAllSpecs(src, ignoreSrc);
  const pkgNames = Object.keys(allSpecs);

  // Update all package.json files with this version
  pkgNames.forEach(pkgName => {
    const { specPath, specs: prevSpecs } = allSpecs[pkgName];
    let nextSpecs = prevSpecs;
    deps.forEach(dep => {
      const { name: depName } = parseDep(dep);
      DEP_TYPES.forEach(type => {
        const depsOfType = nextSpecs[type] || {};
        if (depsOfType[depName] != null) {
          const nextDeps = omit(depsOfType, [depName]);
          nextSpecs = timmSet(nextSpecs, type, nextDeps);
        }
      });
    });
    if (nextSpecs !== prevSpecs) writeSpecs(specPath, nextSpecs);
  });
};

export default run;
