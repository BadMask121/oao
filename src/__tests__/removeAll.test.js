/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import removeAll from '../removeAll';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

describe('REMOVE-ALL command', () => {
  it("leaves package.json untouched if it doesn't depend on dep", async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await removeAll(['mady'], {
      src: 'test/fixtures/packages/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(0);
  });

  it('modifies package.json if it depends on dep', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await removeAll(['ext-one'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();
  });

  it('removes a dep everywhere it may appear', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await removeAll(['ext-two'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(1);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
  });

  it('removes multiple deps', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await removeAll(['ext-one', 'ext-two'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();
  });
});
