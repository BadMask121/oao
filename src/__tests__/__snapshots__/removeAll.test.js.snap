// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REMOVE-ALL command modifies package.json if it depends on dep 1`] = `
Object {
  "dependencies": Object {
    "oao": "*",
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;

exports[`REMOVE-ALL command modifies package.json if it depends on dep 2`] = `
Object {
  "dependencies": Object {
    "oao": "*",
    "timm": "1.x",
  },
  "devDependencies": Object {
    "ext-two": "*",
    "oao-b": "*",
    "xxl": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-c",
  "peerDependencies": Object {
    "ext-two": "*",
    "oao-b": "*",
  },
  "version": "0.1.0",
}
`;

exports[`REMOVE-ALL command removes a dep everywhere it may appear 1`] = `
Object {
  "dependencies": Object {
    "ext-one": "*",
    "oao": "*",
    "timm": "1.x",
  },
  "devDependencies": Object {
    "oao-b": "*",
    "xxl": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-c",
  "peerDependencies": Object {
    "oao-b": "*",
  },
  "version": "0.1.0",
}
`;

exports[`REMOVE-ALL command removes multiple deps 1`] = `
Object {
  "dependencies": Object {
    "oao": "*",
    "timm": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-b",
  "version": "0.1.0",
}
`;

exports[`REMOVE-ALL command removes multiple deps 2`] = `
Object {
  "dependencies": Object {
    "oao": "*",
    "timm": "1.x",
  },
  "devDependencies": Object {
    "oao-b": "*",
    "xxl": "1.x",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "oao-c",
  "peerDependencies": Object {
    "oao-b": "*",
  },
  "version": "0.1.0",
}
`;
